src/plugins/data/public/index_patterns/fields/index_pattern_field.ts

```diff
@@ -0,0 +1,182 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { IndexPattern } from '../index_patterns';
+
+import {
+  IFieldType,
+  IFieldSubType,
+  KbnFieldType,
+  getKbnFieldType,
+  KBN_FIELD_TYPES,
+} from '../../../common';
+
+export interface FieldSpec {
+  count: number;
+  script?: string;
+  lang?: string;
+  conflictDescriptions?: Record<string, string[]>;
+
+  name: string;
+  type: string;
+  esTypes?: string[];
+  scripted: boolean;
+  searchable: boolean;
+  aggregatable: boolean;
+  readFromDocValues?: boolean;
+  subType?: IFieldSubType;
+  indexed?: boolean;
+}
+
+export type OnUnknownType = (unknownType: {
+  type: string;
+  name: string;
+  indexPatternTitle: string;
+}) => void;
+
+export class IndexPatternField implements IFieldType {
+  readonly spec: FieldSpec;
+  // not writable or serialized
+  readonly indexPattern: IndexPattern;
+  readonly displayName: string;
+  private readonly kbnFieldType: KbnFieldType;
+
+  constructor(
+    indexPattern: IndexPattern,
+    spec: FieldSpec,
+    displayName: string,
+    onUnknownType: OnUnknownType
+  ) {
+    this.indexPattern = indexPattern;
+    this.spec = { ...spec, type: spec.name === '_source' ? '_source' : spec.type };
+    this.displayName = displayName;
+
+    this.kbnFieldType = getKbnFieldType(spec.type);
```

Comment:
try to refactor this code so it can provide a return value which can be handled.